### PR TITLE
Add support for API serach by labelName attribute

### PR DESCRIPTION
**Pull request by msm1992**
### Proposed changes in this pull request

This PR contains publisher REST api changes to support api search by labelName attribute.

Other modifications to /apis GET resource

1. a parameter ‘expand’  was introduced to specify whether the returned response should contain APIDTOs or APIInfoDTOs.
Expand, true → APIDTO 
* APIDTO should be changed to APIDetailDTO, that change is not included in this PR for the moment.

2. added support for “Accept: application/gzip” header

sample curl to retrive API details of APIs under label1 as a gzip.

curl -k -H "Authorization: Bearer <access_token>" “https://localhost:9443/api/am/publisher/v0.12/apis?query=labelName:label1&expand=true” -H “Accept: application/gzip” > test.json.gz

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/GZIPUtils.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    public static File constructZippedResponse(Object data) throws APIManagementException {
+        String tmpSourceFileName = System.currentTimeMillis() + ".json";
+        String tmpDestinationFileName = System.currentTimeMillis() + ".json.gz";
+        String sourcePath = System.getProperty(APIConstants.JAVA_IO_TMPDIR) + File.separator  + tmpSourceFileName;
+        String destinationPath = System.getProperty(APIConstants.JAVA_IO_TMPDIR) + File.separator + tmpDestinationFileName;
+        File zippedResponse;
+
+        File file = new File(sourcePath);
+        FileWriter fileWriter;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/GZIPUtils.java]: should we close file writer?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/GZIPUtils.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            gzipOutputStream = new GZIPOutputStream(fileOutputStream);
+            fileInputStream = new FileInputStream(sourcePath);
+
+            int length;
+            while ((length = fileInputStream.read(buffer)) > 0) {
+                gzipOutputStream.write(buffer, 0, length);
+            }
+            gzipOutputStream.finish();
+        } catch (IOException e) {
+            throw new APIManagementException("Error while compressing file at" + sourcePath + "to" + destinationPath, e);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/GZIPUtils.java]: lets put space after file at to

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/GZIPUtils.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        } finally {
+            IOUtils.closeQuietly(fileInputStream);
+            IOUtils.closeQuietly(fileOutputStream);
+            IOUtils.closeQuietly(gzipOutputStream);
+        }
+    }
+
+    public static File constructZippedResponse(Object data) throws APIManagementException {
+        String tmpSourceFileName = System.currentTimeMillis() + ".json";
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/GZIPUtils.java]: Lets move ".json" to a String constant.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/GZIPUtils.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        String tmpDestinationFileName = System.currentTimeMillis() + ".json.gz";
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/GZIPUtils.java]: Same here.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/GZIPUtils.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        GZIPOutputStream gzipOutputStream = null;
+        FileOutputStream fileOutputStream = null;
+        FileInputStream fileInputStream = null;
+
+        try {
+            fileOutputStream = new FileOutputStream(destinationPath);
+            gzipOutputStream = new GZIPOutputStream(fileOutputStream);
+            fileInputStream = new FileInputStream(sourcePath);
+
+            int length;
```
Comment (Krishanx92) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/GZIPUtils.java]: Shall we initialize this int lenght = 0;

**Comment by Krishanx92** [comment]
Fix https://github.com/wso2/product-apim/issues/3325

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java [hunk 6]
```diff
@@ -1908,7 +1908,17 @@ public boolean isDuplicateContextTemplate(String contextTemplate) throws APIMana
             // If the query does not contains "=" then it is an errornous scenario.
             if (query.contains("=")) {
                 String[] searchKeys = query.split("=");
-                if (!Arrays.asList(APIConstants.API_SEARCH_PREFIXES).contains(searchKeys[0].toLowerCase())) {
+
+                //prevent api-meta. getting prefixed to labelName and restrict label serach to exact match only
+                if (APIConstants.LABELNAME.equals(searchKeys[0])) {
+                    searchKeys[0] = APIConstants.LABEL_PREFIX + searchKeys[0];
+                    if (searchKeys[1].startsWith("*")) {
```
Comment (nuwand) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: There's a possible array index out of bonds issue here, if the searchKeys array does not have 2 or more items.

Comment (msm1992) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/5551

**Pull request merged**